Домашнее задание 1 и 2

Eduard1507 opened the pull request:
Домашнее задание 1 по созданию фамильного дерева и домашнее задание 2 по сохранению в файл

Eduard1507 committed:
Домашнее задание 1 и 2

Liberate520 commented on src/Person.java:
```diff
@@ -0,0 +1,62 @@
+    private static final long serialVersionUID = 1L;
+
+    private String name;
+    private int age;
```
неверно. Поле возраста не будет автоматически менять с ходом времени. Правильно сделать дату рождения, а возраст высчитывать в методе

Liberate520 commented on src/FileFamilyTreeOperations.java:
```diff
@@ -0,0 +1,29 @@
+class FileFamilyTreeOperations implements FileOperations {
+
+    @Override
+    public void saveToFile(String filename, List<Person> members) {
```
Правильней передавать не список людей, а сразу все дерево, объект дерева. Сделать это конечно сложно, так как у вас именно класс дерева выполняет операции с файловой системой, что будет нарушать принцип единственной ответственности. Сделайте два класса независимыми и используйте оба пока из main

Eduard1507 committed:
Домашнее задание 1 и 2 после изменения и замечаний от преподавателя

Eduard1507 committed:
Домашнее задание 3

Eduard1507 committed:
Домашнее задание 3

Eduard1507 committed:
Домашнее задание 3

Eduard1507 committed:
Домашнее задание 3

Eduard1507 committed:
ДЗ 4 ООП: Обобщения. ч1

Liberate520 commented on src/com/example/familytree/FamilyTree.java:
```diff
@@ -0,0 +1,71 @@
+import java.util.List;
+
+// Класс FamilyTree представляет генеалогическое древо и реализует Serializable и Iterable<T>
+public class FamilyTree<T extends Person> implements Serializable, Iterable<T> {
```
Зависимость параметра от Person будет ошибкой. Если мы захотим использовать класс для составления родословной собак, то получается класс собаки должен будет наследоваться от Person, что конечно же неверно

Eduard1507 committed:
ДЗ 5 и 6 ООП:
- Рефакторинг проекта с семейным деревом с учетом принципов SOLID
- Реализовать паттерн MVP в вашем проекте с семейным деревом

Eduard1507 committed:
ДЗ 5 и 6 ООП:
- Рефакторинг проекта с семейным деревом с учетом принципов SOLID
- Реализовать паттерн MVP в вашем проекте с семейным деревом (редакция)

Liberate520 commented on src/com/example/familytree/main/Main.java:
```diff
@@ -0,0 +1,94 @@
+        boolean running = true;
+
+        while (running) {
+            System.out.println("\nВыберите команду:");
+            System.out.println("1. Добавление человека");
+            System.out.println("2. Сортировка по имени");
+            System.out.println("3. Сортировка по дате рождения");
+            System.out.println("4. Демонстрация фамильного дерева");
```
Работа должна быть в одном из классов. main не должен содержать бизнес логику приложения

Liberate520 replied:
также работа с меню таким образом нарушает второй принцип солид

Liberate520 commented on src/com/example/familytree/mvp/FamilyTreePresenterImpl.java:
```diff
@@ -0,0 +1,108 @@
+        RelationManagerPresenter {
+
+    private final FamilyTreeView view;
+    private FamilyTree<Person> familyTree;
+    private final FileOperations fileOps;
+    private static final DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd-MM-yyyy");
```
Нужен класс Сервис для объединения работы модели

Liberate520 commented on src/com/example/familytree/mvp/FamilyTreePresenterImpl.java:
```diff
@@ -0,0 +1,108 @@
+    public void addPerson(int id, String name, String birthDateStr, String gender) {
+        try {
+            LocalDate birthDate = LocalDate.parse(birthDateStr, formatter);
+            Person person = new Person(id, name, birthDate, gender);
```
презентер не должен создавать людей. Это нарушает первый принцип солид

Eduard1507 committed:
ДЗ 7 Попытался исправить все замечания

Eduard1507 committed:
ДЗ 7 с исправлениями

Liberate520 commented on src/com/example/familytree/service/FamilyTreeService.java:
```diff
@@ -0,0 +1,73 @@
+
+    @Override
+    public void sortByName() {
+        familyTree.sortBy(new NameComparator());
```
Методы сортировки должны быть реализованы в дереве. Нет необходимости передавать компараторы

Liberate520 commented on src/com/example/familytree/commands/CommandHandler.java:
```diff
@@ -0,0 +1,54 @@
+import java.util.Scanner;
+
+public class CommandHandler {
+    private final FamilyTreePresenterImpl presenter;
+    private final FamilyTreeView view;
+    private final Scanner scanner;
```
Очень похож на божественный класс. Он и с презентером взаимодействует и с деревом (которое является частью модели) и со сканером, который используется для взаимодействия с пользователем. Скорей всего здесь нарушается первый принцип солид

Liberate520 commented on src/com/example/familytree/commands/CommandHandler.java:
```diff
@@ -0,0 +1,54 @@
+    private final Map<Integer, Command> commands = new HashMap<>();
+
+    public CommandHandler(FamilyTreePresenter presenter, FamilyTreeView view) {
+        this.presenter = (FamilyTreePresenterImpl) presenter;
```
Чего вдруг используется приведение типа? Почему тогда сразу не попросить FamilyTreePresenterImpl? Ведь если будет передано что то другое, то приложение упадет

Liberate520 commented on src/com/example/familytree/commands/CommandHandler.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    private void displayMenu() {
+        System.out.println("\nВыберите номер команды:");
```
О) еще и сам выводит сообщения в консоль, без участия view) ну точно нарушение первого принципа солид

Liberate520 commented on src/com/example/familytree/commands/FamilyTreeView.java:
```diff
@@ -0,0 +1,4 @@
+package com.example.familytree.commands;
+
+public class FamilyTreeView {
```
Что это и зачем? А самое интересное почему в пакете команд?

Liberate520 commented on src/com/example/familytree/commands/LoadTreeFromFileCommand.java:
```diff
@@ -0,0 +1,26 @@
+
+    @Override
+    public void execute() {
+        System.out.print("Введите имя файла: ");
```
класс команды не должен взаимодействовать с консолью, для этого есть view
